[Enhancement] Bridge java.util.logging to slf4j to fix log leakage

arvindKandpal-ksolves opened the pull request:
## Why I'm doing:

Currently, third-party libraries that rely on `java.util.logging` (JUL), such as `jprotobuf-rpc-socket`, print logs directly to `stdout` (which redirects to `fe.out`) instead of the standard log files. This happens because the JUL-to-SLF4J bridge is not properly initialized, causing log leakage and making debugging difficult as these logs are missing from `fe.log`.

## What I'm doing:

I have implemented the standard bridging mechanism to route `java.util.logging` calls to SLF4J (Log4j2).

Changes:
1.  Added the `org.slf4j:jul-to-slf4j` dependency in `fe/pom.xml` and `fe/fe-core/pom.xml`.
2.  Initialized `LogManager.getLogManager().reset()` and `SLF4JBridgeHandler.install()` in `Log4jConfig.java` to redirect logs.
3.  Added a unit test `testJulToSlf4jBridge` in `Log4jConfigTest.java` to verify that JUL logs are correctly captured by Log4j2.

This PR supersedes and fixes the unmerged PR #56156.

Fixes #31777

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc. (Logging behavior improvement)

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Ensures logs from JUL-based libraries flow into the existing SLF4J/Log4j2 pipeline and match configured log levels.
> 
> - Add `org.slf4j:jul-to-slf4j` dependency in `fe/pom.xml` and `fe/fe-core/pom.xml`
> - In `Log4jConfig.reconfig()`, install `SLF4JBridgeHandler`, remove root JUL handlers, and set JUL root level based on `sys_log_level` (`DEBUG`→`FINE`, `INFO`→`INFO`, `WARN`→`WARNING`, `ERROR|FATAL`→`SEVERE`)
> - Add unit tests in `Log4jConfigTest`: `testJulToSlf4jBridge` (JUL logs captured by Log4j2) and `testJulLevelMapping` (JUL level mapping and invalid level handling)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d1aaf5e477b8b878700da7b04372a80f7c168155. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

kevincai commented on fe/fe-core/src/main/java/com/starrocks/common/Log4jConfig.java:
```diff
@@ -335,6 +337,8 @@ static String generateActiveLog4jXmlConfig() throws IOException {
 
     private static void reconfig() throws IOException {
+        LogManager.getLogManager().reset();
+        SLF4JBridgeHandler.install();
```
No. if set the level to ALL, the bridge will be all flooded and the perf wil be greatly impacted. https://www.slf4j.org/legacy.html#jul-to-slf4j

arvindKandpal-ksolves committed:
[Enhancement] Bridge java.util.logging to slf4j

Fixes #31777.

This PR bridges java.util.logging logs to slf4j so they are handled by log4j2
configuration instead of being leaked to stdout. This is necessary for
libraries like jprotobuf-rpc-socket which use java.util.logging.

Changes:
1. Add jul-to-slf4j dependency.
2. Initialize SLF4JBridgeHandler in Log4jConfig with dynamic level alignment.
3. Replace LogManager.reset() with removeHandlersForRootLogger() for safety.
4. Add unit test to verify the bridge.

Signed-off-by: arvindksi274-ksolves <arvind.kandpal@ksolves.com>

arvindKandpal-ksolves committed:
[Enhancement] Bridge java.util.logging to slf4j

Fixes #31777.

This PR bridges java.util.logging logs to slf4j so they are handled by log4j2
configuration instead of being leaked to stdout. This is necessary for
libraries like jprotobuf-rpc-socket which use java.util.logging.

Changes:
1. Add jul-to-slf4j dependency.
2. Initialize SLF4JBridgeHandler in Log4jConfig with dynamic level alignment.
3. Replace LogManager.reset() with removeHandlersForRootLogger() for safety.
4. Add validation logic to default to INFO for unsupported levels.
5. Add unit test to verify the bridge, level mapping, and fallback logic.

Signed-off-by: arvindksi274-ksolves <arvind.kandpal@ksolves.com>

arvindKandpal-ksolves commented:
@kevincai I have pushed the final implementation addressing all feedback.

### 1. Regarding Logic in `reconfig()` (Addressing Copilot)
I noticed Copilot suggested using `LogManager.getLogManager().reset()`. However, I intentionally decided against it.
- **Reason:** `reset()` is a destructive operation that closes **all** handlers across the entire JVM logging context. This is risky as it could break logging for other third-party components (like Hadoop/Netty) running in the same process.
- **Solution:** I used `SLF4JBridgeHandler.removeHandlersForRootLogger()`. This is a surgical approach that strictly prevents the stdout leakage (our main goal) without wiping out the global logging state.

### 2. Implementation Updates
- **Dynamic Level Alignment:** JUL levels now sync with StarRocks (`sys_log_level`) to ensure performance.
- **Validation:** Added strict validation using `DEBUG_LEVELS.contains()` logic (defaulting to INFO for unknown levels like TRACE).
- **Test Isolation:** Updated `testJulLevelMapping` to verify the validation logic and added a `try-finally` block to strictly restore the original log levels.

### 3. CI Failure
I noticed the `SQL-Tester` failed with `Base Compaction Failing` and `Manual compaction task is running`. Since this PR only modifies Logging Configuration and has no impact on Backend storage/compaction logic, this appears to be a flaky test/environment issue.

Could you please trigger the CI checks again?

kevincai commented:
> @kevincai Thanks for the approval and the rebase!
> 
> I noticed the `SQL-Tester` failed with `couldn't get JNIEnv` (BE:10003) and `Backend node not found`.
> 
> Since this PR touches the Java logging initialization, could this be a side effect causing the JVM/BE to crash, or does this look like a known flaky CI issue to you?
> 
> Please let me know if you want me to revert anything.

It's an CI env issue, I will take care of it.

arvindKandpal-ksolves commented:
@kevincai , I noticed the AI review (Cursor Bot) raised two valid logical points that I didn't want to leave unresolved. I have pushed a small follow-up fix:

1. **Atomic Configuration:** Moved `generateActiveLog4jXmlConfig()` to the start of `reconfig()`. This ensures validation happens *before* modifying any JUL state, preventing inconsistent states if the config is invalid.
2. **Strict Testing:** Added `Assertions.fail()` in the test case to strictly ensure the test fails if the expected exception isn't thrown.

This makes the implementation safer and the tests more robust. Ready for final review!
